File: src/divesites/dayislandwall.erl
%%******************************************************************************************************************
%%*******************************************************************************************************************
% DivePredictor - A Erlang engine to compute suitability to dive a site, based on prolog-like rules, and
% cached tide/current data from the NOAA.
%
%   Copyright (C) 2014  Archis Gore
%    This file is part of DivePredictor.
%
%    DivePredictor is free software: you can redistribute it and/or modify
%    it under the terms of the GNU Affero General Public License as
%    published by the Free Software Foundation, either version 3 of the
%    License, or (at your option) any later version.
%
%    DivePredictor is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU Affero General Public License for more details.
%
%    You should have received a copy of the GNU Affero General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.
%%******************************************************************************************************************
%%*******************************************************************************************************************/

-module(dayislandwall).
-export([site_info/0]).
-include("include/divepredictor.hrl").


site_info() -> 
	#divesite{
		id="dayislandwall", 
		name="Day Island Wall",
		noaaTideStationId="9447110",
		noaaCurrentStationId="PUG1528",
		location=#divelocation{address="W Day Island Boulevard West and E Day Island Blvd W, Tacoma, Pierce, Washington 98466, United States"},
		find_solutions=fun find_solutions/2
	}.

find_solutions(_, Currents) ->
	io:fwrite("Solutions Finder for Day Island Wall called.~n"),
	[#divesolution{siteId="dayislandwall",time=correct_time(Slack#current.dateTime), length=-1, 
		description=io_lib:format("~p Exchange across ~p minutes: Slack at ~s between ~s(~p)@~s - ~s(~p)@~s", 
				[abs(Before#current.magnitude) + abs(After#current.magnitude),
				((calendar:datetime_to_gregorian_seconds(After#current.dateTime) - calendar:datetime_to_gregorian_seconds(Before#current.dateTime)) / 60),
				divepredictorformatting:datetime_to_string(Slack#current.dateTime), 
				Before#current.type, Before#current.magnitude, divepredictorformatting:datetime_to_string(Before#current.dateTime),
				After#current.type, After#current.magnitude, divepredictorformatting:datetime_to_string(After#current.dateTime)])} 
		||	{Before, Slack, After} <- get_safe_slacks(Currents)].


correct_time(DateTime) ->
	calendar:gregorian_seconds_to_datetime((calendar:datetime_to_gregorian_seconds(DateTime) - 3600)).

get_safe_slacks([]) -> [];
get_safe_slacks([First|Rest]) -> get_safe_slacks(First, Rest).

get_safe_slacks(_, []) -> [];
get_safe_slacks(First, [Second| Rest]) -> get_safe_slacks(First, Second, Rest).

get_safe_slacks(_, _, []) -> [];
get_safe_slacks(First, Second, [Third | Rest]) -> 
	lists:flatten([get_possible_solution(First, Second, Third), get_safe_slacks(lists:flatten([Second, Third, Rest]))]).


get_possible_solution(First,Second,Third) -> 
	case Second#current.type of
		"slack" -> 
			%io:fwrite("Asked to test ~p < ~p < ~p and more~n", [First, Second, Third]),
			case is_safe(First, Third) of
				true -> [{First, Second, Third}];
				_ -> [] end;
		_ -> [] end.

is_safe(First, Third) ->
	FirstMag = abs(First#current.magnitude),
	SecondMag = abs(Third#current.magnitude),
	%Exchange = Flood + Ebb,
	%Time = (calendar:datetime_to_gregorian_seconds(Third#current.dateTime) - calendar:datetime_to_gregorian_seconds(First#current.dateTime)) / 360.00,
	%io:fwrite("Exchange between ~p(~p)-~p(~p) is ~p (Safe:~p)", [First#current.type, First#current.magnitude,
	%		Third#current.type, Third#current.magnitude, Exchange, Exchange =< 5]),
	(FirstMag < 3) and (SecondMag < 3) and ((FirstMag + SecondMag) < 4).

File: python/sites/dayislandwall.py
from datetime import timedelta

from models import Current, DiveLocation, DiveSite, DiveSolution, Tide
from sites._common import fmt, safe_slack_triples

# Day Island Wall sits in the side passage west of the Tacoma Narrows.
# The closest NOAA current station, PUG1528, is "The Narrows, South end
# (midstream)" — in the main channel, around the corner from the wall.
# In Puget Sound's geometry, slack in a side passage typically leads or
# lags the main channel by 30-90 min; empirically, slack at the wall
# occurs about an hour earlier than PUG1528's prediction. We subtract
# that here so displayed times reflect the wall, not the midstream
# Narrows station.
LOCAL_SLACK_OFFSET = timedelta(hours=-1)

site = DiveSite(
    id="dayislandwall",
    name="Day Island Wall",
    tide_station_id="9447110",
    current_station_id="PUG1528",
    location=DiveLocation(
        address="W Day Island Boulevard West and E Day Island Blvd W, Tacoma, Pierce, Washington 98466, United States"
    ),
    notes=(
        "Spatial slack offset: the nearest NOAA current station, PUG1528 "
        "(\"The Narrows, South end, midstream\"), sits in the main channel "
        "around the corner from the wall. Empirically, slack at Day Island "
        "Wall happens ~1 h before the NOAA prediction at the Narrows. "
        "Listed times already apply that correction — dive these as the "
        "slack itself."
    ),
)


def _keep(before: Current, slack: Current, after: Current) -> bool:
    if slack.type != "slack":
        return False
    first_mag = abs(before.magnitude)
    second_mag = abs(after.magnitude)
    return (first_mag < 3) and (second_mag < 3) and ((first_mag + second_mag) < 4)


def find_solutions(_tides: list[Tide], currents: list[Current]) -> list[DiveSolution]:
    out: list[DiveSolution] = []
    for b, slack, a in safe_slack_triples(currents, _keep):
        local_slack = slack.dt + LOCAL_SLACK_OFFSET
        exch = abs(b.magnitude) + abs(a.magnitude)
        minutes = (a.dt - b.dt).total_seconds() / 60.0
        desc = (
            f"Estimated slack at the wall (NOAA Narrows prediction "
            f"{fmt(slack.dt)} − 1 h). "
            f"{exch:.2f} kt exchange across {minutes:.0f} min: "
            f"{b.type}({b.magnitude})@{fmt(b.dt)} → {a.type}({a.magnitude})@{fmt(a.dt)}."
        )
        out.append(DiveSolution(site_id=site.id, time=local_slack, description=desc))
    return out
